server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/NodeReplacementAllocationDecider.java

```diff
@@ -0,0 +1,155 @@
+        }
+    }
+
+    @Override
+    public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (replacementOngoing(allocation) == false) {
+            return NO_REPLACEMENTS;
+        } else if (isReplacementSource(allocation, node.nodeId())) {
+            return Decision.single(Decision.Type.NO, NAME,
+                "node [%s] is being replaced by node [%s], so no data may remain on it", node.nodeId(),
+                getReplacementName(allocation, node.nodeId()));
+        } else {
+            return Decision.single(Decision.Type.YES, NAME, "node [%s] is not being replaced", node.nodeId());
+        }
+    }
+
+    @Override
+    public Decision shouldAutoExpandToNode(IndexMetadata indexMetadata, DiscoveryNode node, RoutingAllocation allocation) {
+        if (replacementOngoing(allocation) == false) {
+            return NO_REPLACEMENTS;
+        } else if (isReplacementTargetName(allocation, node.getName())) {
+            final SingleNodeShutdownMetadata shutdown = allocation.replacementTargetShutdowns().get(node.getName());
+            return Decision.single(Decision.Type.NO, NAME,
+                "node [%s] is a node replacement target for node [%s], " +
+                    "shards cannot auto expand to be on it until the replacement is complete",
+                node.getId(), shutdown == null ? null : shutdown.getNodeId());
+        } else if (isReplacementSource(allocation, node.getId())) {
+            return Decision.single(Decision.Type.NO, NAME,
+                "node [%s] is being replaced by [%s], shards cannot auto expand to be on it",
+                node.getId(), getReplacementName(allocation, node.getId()));
+        } else {
+            return Decision.single(Decision.Type.YES, NAME,
+                "node is not part of a node replacement, so shards may be auto expanded onto it");
+        }
+    }
+
+    @Override
+    public Decision canForceAllocateDuringReplace(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (replacementFromSourceToTarget(allocation, shardRouting.currentNodeId(), node.node().getName())) {
+            return Decision.single(Decision.Type.YES, NAME,
+                "node [%s] is being replaced by node [%s], and can be force vacated to the target",
+                shardRouting.currentNodeId(), node.nodeId());
+        } else {
+            return Decision.single(Decision.Type.NO, NAME,
+                "shard is not on the source of a node replacement relocated to the replacement target");
+        }
+    }
+
+    /**
+     * Returns true if there are any node replacements ongoing in the cluster
+     */
+    private static boolean replacementOngoing(RoutingAllocation allocation) {
+        return allocation.nodeShutdowns().values().stream()
+            .anyMatch(shutdown -> shutdown.getType().equals(SingleNodeShutdownMetadata.Type.REPLACE));
```

Comment:
I think we can simplify to:
```suggestion
        return allocation.replacementTargetShutdowns.isEmpty() == false;
```